_drafts/writing-a-bootloader-from-scratch.md

```diff
@@ -0,0 +1,630 @@
+            |     Application     |
+            |                     |
+            |                     |
+    0x30000 +---------------------+
+```
+
+We can transcribe that memory into a linker script:
+
+``` 
+/* memory_map.ld */
+MEMORY
+{
+  bootrom  (rx)  : ORIGIN = 0x00000000, LENGTH = 0x00004000
+  approm   (rx)  : ORIGIN = 0x00004000, LENGTH = 0x0003C000
+  ram      (rwx) : ORIGIN = 0x20000000, LENGTH = 0x00008000
+}
+
+__bootrom_start__ = ORIGIN(bootrom);
+__bootrom_size__ = LENGTH(bootrom);
+__approm_start__ = ORIGIN(approm);
+__approm_size__ = LENGTH(approm);
+```
+
+Since linker scripts are composable, we will be able to `include` that memory
+map into the linker scripts we write for our bootloader and our application.
+
+You’ll notice that the linker script above declares some variables. We’ll need
+those for our bootloader to know where to find the application. To make them
+accessible in C code, we declare them in a header file:
+
+```c
+/* memory_map.h */
+#pragma once
+
+extern int __bootrom_start__;
+extern int __bootrom_size__;
+extern int __approm_start__;
+extern int __approm_size__;
+```
+
+### Implementing the bootloader itself
+Next up, let’s write some bootloader code. Our bootloader needs to start
+executing on boot and then jump to our app.
+
+We know how to do the first part from our previous post: we need a valid stack
+pointer at address `0x0` , and a valid `Reset_Handler`  function setting up our
+environment at address `0x4`. We can reuse our previous startup file and linker
+script, with one change: we use  `memory_map.ld` rather than define our own
+`MEMORY` section.
+
+We also need to put our code in the `bootrom` region from our memory rather than
+the `rom` region in our previous post.
+
+Our linker script therefore looks like this:
+
+```
+/* bootloader.ld */
+INCLUDE memory_map.ld
+
+/* Section Definitions */
+SECTIONS
+{
+    .text :
+    {
+        KEEP(*(.vectors .vectors.*))
+        *(.text*)
+        *(.rodata*)
+        _etext = .;
+    } > bootrom
+  ... 
+}
+```
+
+To jump into our application, we need to know where the `Reset_Handler` of the
+app is, and what stack pointer to load.  Again, we know from our previous post
+that those should be the first two 32-bit words in our binary, so we just need
+to dereference those addresses using the `__approm_start__` variable from our
+memory map.
+
+```c
+/* bootloader.c */
+#include <inttypes.h>
+#include "memory_map.h"
+
+int main(void) {
+  uint32_t *app_code = (uint32_t *)__approm_start__;
+  uint32_t app_sp = app_code[0];
+  uint32_t app_start = app_code[1];
+  /* TODO: Start app */
+  /* Not Reached */
+  while (1) {}
+}
+```
+
+Next we must load that stack pointer and jump to the code. This will require a
+bit of assembly code.
+
+ARM MCUs use the [`msr` instruction
+](http://www.keil.com/support/man/docs/armasm/armasm_dom1361289882044.htm) to
+load immediate or register data into system registers, in this case the MSP
```

Comment:
'load immediately'? (not sure)

Reply:
no, load immediate. An immediate is a constant value in assembly. 

>In assembly language, literal numbers and characters are done using the "immediate mode" instructions available on most microprocessors. The name "immediate" comes from the values being available immediately from the instruction stream, as opposed to loading them indirectly by looking up a memory address.[1] On the other hand, values longer than the microprocessor's word length, such as strings and arrays, are handled indirectly and assemblers generally provide a "data" pseudo-op to embed such data tables in a program.

https://en.wikipedia.org/wiki/Constant_(computer_programming)